Refactor the data flow inside of Scaler

evancharlton opened the pull request:
Clean up the internal data flow inside of the Scaler component. This
refactoring had a few main goals:

  1. Fully-own the responsibility of providing scaled data downstream
     alongside the Series and Collection items.
  2. Increase performance by using better checks to detect when the
     incoming Series and Collections objects were externally changed.
  3. Stop exposing `domainsByItemId` / `subDomainsByItemId` to the
     downstream rendering components, in deference to Series and
     Collection objects directly.
  4. Get closer toward enabling per-Series `timeDomains` / `timeSubDomains`
     as documented in #301.

The end result of this is that the (internal) API has nontrivially
changed. Instead of Scaler exposing maps of `domainsByItemId` and
`subDomainByItemId`, it will simply pass through the `series` and
`collections` arrays, but they are now all guaranteed to be
fully-populated with domain and subDomain information. This means that
rendering components can simply render using the `timeDomain`,
`timeSubDomain`, `xDomain`, `xSubDomain`, `yDomain`, and `ySubDomain` domains
(which are guaranteed to be present) and don't need to concern
themselves with how they were populated (user/state, props, calculated
from data, or placeholders).

This does not yet fully enable separate `timeDomain` / `timeSubdomain`
throughout the library, but it removes almost all blockers. The final
piece to this puzzle is in `DataProvider` -- it needs to learn a one more
trick in order to support this (it currently only knows about one
`timeSubDomain`). That cleanup will be saved for a future PR.

cognite-cicd commented:
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-313.surge.sh

cognite-cicd commented:
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-313.surge.sh

cognite-cicd commented:
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-313.surge.sh

cognite-cicd commented:
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-313.surge.sh

valerii-cognite commented on src/components/Griff/index.tsx:
```diff
@@ -0,0 +1,444 @@
+  series: IncomingSeries[];
+  collections: IncomingCollection[];
+
+  children: JSX.Element | JSX.Element[];
```
Can't find `onFetchData` prop, has it been removed in older versions?

valerii-cognite commented on src/components/Griff/index.tsx:
```diff
@@ -0,0 +1,444 @@
+
+    return (
+      <Scaler
+        timeSubDomainChanged={() => null}
```
I probably missed that but I can't find on how it is used

valerii-cognite commented on src/utils/domains.ts:
```diff
@@ -0,0 +1,162 @@
+    return domain;
+  }
+  return newDomain(
+    extent[0] - diff * 0.025,
```
Just wondering what is the reasoning behind `0.025`, `0.25`, `1e-3`?

evancharlton replied:
This was just a buffer that "seemed right". I don't know the reasoning behind it.

valerii-cognite commented on src/utils/placeholder.ts:
```diff
@@ -0,0 +1,28 @@
+export const placeholder = (min: number, max: number): Domain =>
+  newDomain(min, max, DomainPriority.PLACEHOLDER);
+
+export const withoutPlaceholder = (
```
It seems everywhere `withoutPlaceholder` is used we return placeholder as a default value (i.e. `withoutPlaceholder(c.xDomain) || PLACEHOLDER_DOMAIN`), shouldn't it be renamed to something like `withPlaceholder` and take in placeholder as an argument?

valerii-cognite commented on src/components/Griff/index.tsx:
```diff
@@ -0,0 +1,444 @@
+    }, []);
+  };
+
+  registerCollection = ({ id, ...collection }: IncomingCollection) => {
```
Can't `registerCollection` reuse `updateCollection`? Same with `registerSeries`?

evancharlton replied:
Yep, it looks like they can. They were originally different, but that's obsolete now.

valerii-cognite commented on src/components/Griff/index.tsx:
```diff
@@ -0,0 +1,444 @@
+  // Add a helper method to render the legacy props using the new tree structure
+  // format. This is only intended to ease the transition pain and is not
+  // intended to be an ongoing solution.
+  renderLegacyItems = () => {
```
Is this guy obsolete now?

evancharlton replied:
Yeah I think we can remove this since we're making a huge breaking API change anyway.

valerii-cognite commented on src/components/Scaler/index.tsx:
```diff
@@ -1,123 +1,125 @@
 import * as React from 'react';
-import * as PropTypes from 'prop-types';
-import DataContext from '../../context/Data';
```
We should remove unused `../../context/Data`

evancharlton replied:
Good catch!

valerii-cognite commented on src/components/Joiner/index.tsx:
```diff
@@ -0,0 +1,186 @@
+    collectionsById,
+  };
+  return (
+    <GriffContext.Provider value={context}>{children}</GriffContext.Provider>
```
Nice! single source of truth for every component down the tree

valerii-cognite commented on src/components/DataProvider/index.tsx:
```diff
@@ -0,0 +1,199 @@
+  componentDidUpdate() {
+    const { series } = this.props;
+
+    series.forEach((s: ScaledSeries) => {
```
Should we restrict this call only to series changes?

evancharlton replied:
I originally did that, but my thinking was that if a series didn't change, then it'll be a no-op to update it. Filtering them out could only lead to complications (and possibly a small performance boost).

I'm of the opinion that we can add this in later once/if there are performance problems -- what do you think?

valerii-cognite replied:
sure, feel free

valerii-cognite commented on src/components/DataProvider/index.tsx:
```diff
@@ -0,0 +1,199 @@
+    });
+  }
+
+  getLoaderReason = (series: ScaledSeries) => {
```
I don't see `INTERVAL` and `UPDATE_POINTS_PER_SERIES` have they been replaced or obsolete with a newer logic?

valerii-cognite commented on stories/LineChart.stories.js:
```diff
@@ -828,97 +979,12 @@ storiesOf('LineChart', module)
-    <DataProvider
-      defaultLoader={liveLoader}
+    <Griff
+      loader={liveLoader}
```
I believe entire functionality about `isTimeSubDomainSticky` is gone

evancharlton commented on src/components/XAxis/index.tsx:
```diff
@@ -254,30 +265,27 @@ const XAxis: React.FunctionComponent<Props & ScalerProps & SizeProps> = ({
+    {({ size }: { size: SizeProps }) => (
+      <GriffContext.Consumer>
+        {({ series }: ScalerProps) => {
+          console.log(size.width);
```
Whoops; good catch!

evancharlton commented:
Thanks for the review, @valerii-cognite ! I've gone ahead and updated the README and addressed some of the comments -- I'll continue to work on the others.
